src/Common/GuardedPoolAllocator.cpp

```diff
@@ -0,0 +1,328 @@
+    else
+        adjusted_sample_rate_plus_one = 2;
+
+    getThreadLocals()->next_sample_counter = (getRandomUint32T() % (adjusted_sample_rate_plus_one - 1)) + 1;
+    allocator_ptr = this;
+}
+
+void GuardedPoolAllocator::lock()
+{
+    getThreadLocals()->recursive_guard = true;
+    /// Lock mutex to lock any in-progress allocations
+    pool_reservation_mutex.lock();
+}
+
+void * GuardedPoolAllocator::allocate(size_t size, size_t alignment)
+{
+    if (size > state.maximumAllocationSize())
+        return nullptr;
+
+    if (state.guarded_page_pool_end == 0)
+    {
+        getThreadLocals()->next_sample_counter = adjusted_sample_rate_plus_one - 1;
+        return nullptr;
+    }
+
+
+    /// Protect against recursivity.
+    if (getThreadLocals()->recursive_guard)
+        return nullptr;
+    ScopedRecursiveGuard recursive_guard;
+
+    if (alignment == 0)
+        alignment = alignof(max_align_t);
+
+    size_t slot_index;
+    {
+        std::lock_guard guard{pool_reservation_mutex};
+        slot_index = reserveSlotInGuardedPool();
+    }
+
+    if (slot_index == kInvalidSlotID)
+    {
+        return nullptr;
+    }
+
+    uintptr_t slot_start = state.slotToAddr(slot_index);
+    uintptr_t slot_end = state.slotToAddr(slot_index) + state.maximumAllocationSize();
+    AllocationMetadata * meta = addrToMetadata(slot_start);
+
+    uintptr_t user_ptr;
+    /// Randomly choose whether to left-align or right-align the allocation with
+    /// respect to allocated chunk, and then apply the necessary adjustments
+    /// to get an aligned pointer.
+    if (getRandomUint32T() % 2 == 0)
+        user_ptr = alignUp(slot_start, alignment);
+    else
+        user_ptr = alignDown(slot_end - size, alignment);
+
+    assert(user_ptr >= slot_start);
+    assert(user_ptr + size <= slot_end);
+
+    /// If a slot is multiple pages in size, and the allocation takes up a single
+    /// page, we can improve overflow detection by leaving the unused pages as
+    /// unmapped. Thus we do set PROT_READ | PROT_WRITE on entire slot. Instead
+    /// we set those only to pages, which will be used
+    const size_t page_size = state.page_size;
+    allocateInGuardedPool(reinterpret_cast<void *>(getPageAddr(user_ptr, page_size)), roundUpTo(size, page_size));
+
+    meta->recordAllocation(user_ptr, size);
+    meta->allocation_trace.recordBacktrace();
+
+    /// I want to remove this output at the last moment before merge
+    // printf(
+    //     "Given out: \n  slot=%zu\n  address=(%p)\n  (start=%p, end=%p)\n",
+    //     slot_index,
+    //     reinterpret_cast<void *>(user_ptr),
+    //     reinterpret_cast<void *>(user_ptr),
+    //     reinterpret_cast<void *>(user_ptr + roundUpTo(size, page_size)));
+    // printf("user_ptr=%p, slot_end=%p\n", reinterpret_cast<void *>(user_ptr), reinterpret_cast<void *>(slot_end));
+    // printf("Distance to guard page: %zu\n", slot_end - user_ptr);
+
+    // printf("\n");
```

Comment:
Let's also add profile events for guarded allocations, this way we can measure the difference between queries with them and w/o (this will help ensuring that this pool does not affects performance tests for instance).

For this you need to add new events to https://github.com/ClickHouse/ClickHouse/blob/ef35fa6a46ee763c04c17042651b474890cbc1fa/src/Common/ProfileEvents.cpp#L261

```c
    M(GuardedPoolAllocations, "Number of times when allocation was satisfied from guarded pool.") \
    M(GuardedPoolDeallocations, "Number of times when deallocation was from guarded pool.") \
```

And them use them here like (somewhere in the beginning of this file):

```c
#include <Common/ProfileEvents.h>

namespace ProfileEvents
{
    extern const Event GuardedPoolAllocations;
    extern const Event GuardedPoolDeallocations;
}
```

And in this function:

```c
ProfileEvents::increment(ProfileEvents::GuardedPoolAllocations);
```

And don't forget to do the same for deallocations.

Reply:
Done this, now waiting for result.